### PR TITLE
Command parser

### PR DESCRIPTION
**Pull request by clero**
Delegate remote command parsing

ParameterMgr had many responsibilities. For example, it was in charge of 
the remote commands' parsing.
This group of patch is reworking the command parsing handling. The parsing is
now handled by a dedicated class. It avoids to load remote commands parsing structure
if the tuning mode is not allowed.
As test-platform was doing the same work for its commands, a dedicated class has also been
created for it.

### CODE REVIEW

**Comment by clero** [comment]
should be rebased on #94 
replace #54 

@dawagner please review

File: parameter/ParameterMgr.cpp [hunk 1]
```diff
@@ -335,6 +147,7 @@ CParameterMgr::CParameterMgr(const string& strConfigurationFilePath, log::ILogge
     _pvLibRemoteProcessorHandle(NULL),
     _uiStructureChecksum(0),
     _pRemoteProcessorServer(NULL),
+    _commandParser(*this),
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: further enhancement: only instantiate it if/when the remote interface is needed (e.g. don't create it if TuningAllowed=false).

File: parameter/command/include/command/Parser.h [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    /** Internal command handler getter
+     *
+     * @return pointer on the internal command handler
+     */
+    CommandHandler* getCommandHandler();
+
+private:
+
+    /** Command handler method return type */
+    typedef CommandHandler::CommandStatus CommandReturn;
```
Comment (dawagner) [inline parameter/command/include/command/Parser.h]: prefer "using".

File: parameter/command/include/command/Parser.h [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+    CommandReturn importConfigurableDomainsFromXML(const IRemoteCommand& remoteCommand,
+                                                   std::string& result);
+
+    CommandReturn exportConfigurableDomainsWithSettingsToXML(const IRemoteCommand& remoteCommand,
+                                                             std::string& result);
+
+    CommandReturn importConfigurableDomainsWithSettingsFromXML(const IRemoteCommand& remoteCommand,
+                                                               std::string& result);
+
+    CommandReturn exportConfigurableDomainWithSettingsFromXML(const IRemoteCommand& remoteCommand,
```
Comment (dawagner) [inline parameter/command/include/command/Parser.h]: oops, this name should be export...**To**XML. (my bad, I did a mistake during the rebase)

File: parameter/command/include/command/Parser.h (hunk 2 above)
Comment (clero) [inline parameter/command/include/command/Parser.h]: done

File: parameter/command/include/command/Parser.h (hunk 3 above)
Comment (clero) [inline parameter/command/include/command/Parser.h]: done

File: remote-processor/RemoteCommandHandlerTemplate.h [hunk 4]
```diff
@@ -112,144 +137,95 @@ class TRemoteCommandHandlerTemplate : public IRemoteCommandHandler
...
-            return false;
-        }
+        /** Command description string */
+        std::string mDescription;
+    };
 
-        // Add command
-        _remoteCommandParserVector.push_back(new CRemoteCommandParserItem(strCommandName, pfnParser, uiMinArgumentCount, strHelp, strDescription));
+    /** Remote command parser container type */
+    typedef std::map<std::string, RemoteCommandParserItem> RemoteCommandParserItems;
```
Comment (dawagner) [inline remote-processor/RemoteCommandHandlerTemplate.h]: using

File: remote-processor/RemoteCommandHandlerTemplate.h [hunk 5]
```diff
@@ -112,144 +137,95 @@ class TRemoteCommandHandlerTemplate : public IRemoteCommandHandler
...
-    // Parsers
-    bool addCommandParser(const std::string& strCommandName,
-        RemoteCommandParser pfnParser,
-        uint32_t uiMinArgumentCount,
-        const std::string& strHelp,
-        const std::string& strDescription)
-    {
-        if (findCommandParserItem(strCommandName)) {
+        /** Command help string */
+        std::string mHelp;
```
Comment (dawagner) [inline remote-processor/RemoteCommandHandlerTemplate.h]: "help" isn't clear enough about what it is. What it actually contains is the description of the arguments.

File: remote-processor/RemoteCommandHandlerTemplate.h [hunk 6]
```diff
@@ -112,144 +137,95 @@ class TRemoteCommandHandlerTemplate : public IRemoteCommandHandler
...
-
-            while (uiToSpacesAdd--) {
-
-                strResult += " ";
-            }
-
-            strResult += std::string("=> ") + std::string(pRemoteCommandParserItem->getDescription()) + "\n";
+        for (auto& help : helps) {
+            help.usage.resize(maxUsage, ' ');
+            result += help.usage + " => " + help.description + '\n';
```
Comment (dawagner) [inline remote-processor/RemoteCommandHandlerTemplate.h]: Replace "\n" by "std::endl"

File: test/test-platform/CMakeLists.txt [hunk 7]
```diff
@@ -28,12 +28,14 @@
 
 add_executable(test-platform
     main.cpp
-    TestPlatform.cpp)
+    TestPlatform.cpp
+    command/src/Parser.cpp)
 
-# FIXME: Supress the need for the -Wno-unused-parameter
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-parameter")
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
```
Comment (dawagner) [inline test/test-platform/CMakeLists.txt]: no-op. you can remove this line

File: test/test-platform/TestPlatform.h [hunk 8]
```diff
@@ -29,141 +29,59 @@
...
-     *
-     * @param[in] remoteCommand is ignored
-     *
-     * @return EDone (never fails)
-     */
-    CommandReturn applyConfigurations(
-            const IRemoteCommand& remoteCommand, std::string& strResult);
+public:
+    CTestPlatform(const std::string &strclass, int iPortNumber, sem_t& exitSemaphore);
+    virtual ~CTestPlatform();
```
Comment (dawagner) [inline test/test-platform/TestPlatform.h]: shouldn't this class be made final instead?

File: test/test-platform/command/src/Parser.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+    bool bSuccess;
+
+    const char* pcState = remoteCommand.getArgument(1).c_str();
+
+    char* pcStrEnd;
+
+    // Reset errno to check if it is updated during the conversion (strtol/strtoul)
+    errno = 0;
+
+    uint32_t state = strtoul(pcState, &pcStrEnd, 0);
```
Comment (dawagner) [inline test/test-platform/command/src/Parser.cpp]: use convertTo instead. (possibily in a separate patch)

File: remote-processor/RemoteCommandHandlerTemplate.h (hunk 4 above)
Comment (clero) [inline remote-processor/RemoteCommandHandlerTemplate.h]: done

File: remote-processor/RemoteCommandHandlerTemplate.h (hunk 5 above)
Comment (clero) [inline remote-processor/RemoteCommandHandlerTemplate.h]: done

File: test/test-platform/CMakeLists.txt (hunk 7 above)
Comment (clero) [inline test/test-platform/CMakeLists.txt]: done

File: test/test-platform/command/src/Parser.cpp (hunk 9 above)
Comment (clero) [inline test/test-platform/command/src/Parser.cpp]: will disappear after rebase

File: parameter/ParameterMgr.cpp (hunk 1 above)
Comment (clero) [inline parameter/ParameterMgr.cpp]: done

File: parameter/ParameterMgr.cpp [hunk 10]
```diff
@@ -515,6 +310,10 @@ bool CParameterMgr::loadFrameworkConfiguration(string& strError)
     // Log tuning availability
     info() << "Tuning " << (_pfwConfiguration.isTuningAllowed() ? "allowed" : "prohibited");
 
+    if(_pfwConfiguration.isTuningAllowed()) {
+        _commandParser = ParserWrapper{new command::Parser(*this)};
+    }
+
```
Comment (dawagner) [inline parameter/ParameterMgr.cpp]: See https://github.com/clero/parameter-framework/commit/8ad588bd5100e525136c2044514fc76866d01f9d#commitcomment-11721629

(+ coding style: add a space after `if`)

File: test/test-platform/command/include/command/Parser.h [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+{
+
+public:
+
+    /** Remote command Handler type */
+    using CommandHandler = RemoteCommandHandlerTemplate<Parser>;
+
+    /** Class constructor
+     *
+     * @param parameterMgr pointer
```
Comment (dawagner) [inline test/test-platform/command/include/command/Parser.h]: copy/paste mistake ?

Comment (clero) [inline test/test-platform/command/include/command/Parser.h]: done

**Pull request merged**